File: frameworks/dotnet/minimal-apis/UnitTests/CacheTests.cs
namespace UnitTests;

public sealed class CacheTests(MinimalApisApp app) : IClassFixture<MinimalApisApp>
{
    // rb:test cache.small,cache.medium,cache.large
    [Theory]
    [Trait("corpus", "cache.small")]
    [Trait("corpus", "cache.medium")]
    [Trait("corpus", "cache.large")]
    [InlineData("small")]
    [InlineData("medium")]
    [InlineData("large")]
    public async Task A_second_request_is_the_stored_answer(string size)
    {
        HttpClient client = app.CreateClient();

        using HttpResponseMessage first = await client.GetAsync($"/cache/{size}");
        using HttpResponseMessage second = await client.GetAsync($"/cache/{size}");

        await Answer.Is(Expected.Json($"items.{size}.json"), second);
        Assert.Equal(Answer.Serial(first), Answer.Serial(second));
    }

    // rb:test cache.vary_one
    [Fact]
    [Trait("corpus", "cache.vary_one")]
    public async Task One_vary_header_keys_the_store()
    {
        long alpha = await Serial("/cache/vary/one", ("x-rb-tenant", "alpha"));
        long beta = await Serial("/cache/vary/one", ("x-rb-tenant", "beta"));

        Assert.Equal(alpha, await Serial("/cache/vary/one", ("x-rb-tenant", "alpha")));
        Assert.NotEqual(alpha, beta);
    }

    // rb:test cache.vary_many
    [Fact]
    [Trait("corpus", "cache.vary_many")]
    public async Task Each_of_three_vary_headers_keys_the_store()
    {
        (string, string)[] webEuAlpha = [("x-rb-channel", "web"), ("x-rb-region", "eu"), ("x-rb-tenant", "alpha")];
        (string, string)[] webEuBeta = [("x-rb-channel", "web"), ("x-rb-region", "eu"), ("x-rb-tenant", "beta")];

        long first = await Serial("/cache/vary/many", webEuAlpha);

        Assert.Equal(first, await Serial("/cache/vary/many", webEuAlpha));
        Assert.NotEqual(first, await Serial("/cache/vary/many", webEuBeta));
    }

    [Fact]
    public async Task The_answer_says_what_it_varies_on()
    {
        using HttpResponseMessage response = await app.CreateClient().GetAsync("/cache/vary/many");

        Assert.Equal("x-rb-channel, x-rb-region, x-rb-tenant", Answer.Header(response, "vary"));
    }

    private async Task<long> Serial(string path, params (string Name, string Value)[] headers)
    {
        using HttpRequestMessage request = new(HttpMethod.Get, path);
        foreach ((string name, string value) in headers)
        {
            request.Headers.Add(name, value);
        }
        using HttpResponseMessage response = await app.CreateClient().SendAsync(request);
        return Answer.Serial(response);
    }
}

File: frameworks/java/javalin/UnitTests/src/test/java/unittests/CacheTests.java
package unittests;

import static org.junit.jupiter.api.Assertions.assertEquals;
import static org.junit.jupiter.api.Assertions.assertNotEquals;

import io.javalin.testtools.HttpClient;
import io.javalin.testtools.JavalinTest;
import io.javalin.testtools.Response;
import org.junit.jupiter.api.Tag;
import org.junit.jupiter.api.Test;
import org.junit.jupiter.params.ParameterizedTest;
import org.junit.jupiter.params.provider.ValueSource;

class CacheTests extends JavalinApp {

    // rb:test cache.small,cache.medium,cache.large
    @ParameterizedTest
    @Tag("cache.small")
    @Tag("cache.medium")
    @Tag("cache.large")
    @ValueSource(strings = {"small", "medium", "large"})
    void aSecondRequestIsTheStoredAnswer(String size) {
        JavalinTest.test(app(), (server, client) -> {
            long first = Answer.serial(client.get("/cache/" + size));
            Response second = client.get("/cache/" + size);

            Answer.is(Expected.json("items." + size + ".json"), second);
            assertEquals(first, Answer.serial(second));
        });
    }

    // rb:test cache.vary_one
    @Test
    @Tag("cache.vary_one")
    void oneHeaderKeysTheStore() {
        JavalinTest.test(app(), (server, client) -> {
            long alpha = get(client, "/cache/vary/one", "x-rb-tenant", "alpha");
            long beta = get(client, "/cache/vary/one", "x-rb-tenant", "beta");

            assertEquals(alpha, get(client, "/cache/vary/one", "x-rb-tenant", "alpha"));
            assertNotEquals(alpha, beta);
        });
    }

    // rb:test cache.vary_many
    @Test
    @Tag("cache.vary_many")
    void eachOfThreeHeadersKeysTheStore() {
        String[] webEuAlpha = {"x-rb-channel", "web", "x-rb-region", "eu", "x-rb-tenant", "alpha"};
        String[] webEuBeta = {"x-rb-channel", "web", "x-rb-region", "eu", "x-rb-tenant", "beta"};

        JavalinTest.test(app(), (server, client) -> {
            long first = get(client, "/cache/vary/many", webEuAlpha);

            assertEquals(first, get(client, "/cache/vary/many", webEuAlpha));
            assertNotEquals(first, get(client, "/cache/vary/many", webEuBeta));
        });
    }

    @Test
    void theAnswerAndItsReplaySayWhatTheyVaryOn() {
        JavalinTest.test(app(), (server, client) -> {
            assertEquals("x-rb-channel, x-rb-region, x-rb-tenant", Answer.header(client.get("/cache/vary/many"), "vary"));
            assertEquals("x-rb-channel, x-rb-region, x-rb-tenant", Answer.header(client.get("/cache/vary/many"), "vary"));
        });
    }

    /** The serial of the answer to a GET with each pair of strings as a header. */
    private static long get(HttpClient client, String path, String... headers) {
        return Answer.serial(client.get(path, request -> {
            for (int i = 0; i < headers.length; i += 2) {
                request.header(headers[i], headers[i + 1]);
            }
        }));
    }
}
